File: src/main/java/wumo/rl/model/MDP.java
package wumo.rl.model;

import java.util.Iterator;
import java.util.function.Supplier;

/**
 * <p>
 * Created on 2017-07-13.
 * </p>
 *
 * @author wumo
 */
public class MDP {
    public interface GetterSetter<Idx, E> {
        E get(Idx index);
        
        void set(Idx index, E s);
    }
    
    public interface StateSet extends GetterSetter<Object, State>, Iterable<State> {}
    
    public static class State {
        public ActionSet actions;
    }
    
    public interface ActionSet extends GetterSetter<Object, Action>, Iterable<Action> {
        default Action firstOrNull() {
            Iterator<Action> iter = this.iterator();
            if (iter.hasNext()) return iter.next();
            return null;
        }
    }
    
    public static class Action {
        public final String desc;
        public final int idx;
        
        public Action(int idx, String desc) {
            this.idx = idx;
            this.desc = desc;
        }
        
        public PossibleSet possibles;
        
        @Override public String toString() {
            return desc;
        }
    }
    
    public interface PossibleSet extends GetterSetter<Object, Possible>, Iterable<Possible> {}
    
    public static class Possible {
        public State state;
        public double reward;
        public double probability;
        
        public Possible() {
        }
        
        public Possible(State state, double reward, double probability) {
            this.state = state;
            this.reward = reward;
            this.probability = probability;
        }
    }
    
    public interface StateValueFunction extends GetterSetter<State, Double> {}
    
    public interface ActionValueFunction {
        double get(State s, Action a);
        
        void set(State s, Action a, double value);
    }
    
    public interface Policy extends GetterSetter<State, Action> {}
    
    public final StateSet states;
    public final Supplier<? extends StateValueFunction> V_maker;
    public final Supplier<? extends Policy> pi_maker;
    public final Supplier<? extends ActionValueFunction> Q_maker;
    public final double gamma;
    
    public MDP(StateSet states, double gamma, Supplier<? extends StateValueFunction> v_maker, Supplier<? extends Policy> pi_maker,
               Supplier<? extends ActionValueFunction> q_maker) {
        this.states = states;
        V_maker = v_maker;
        this.pi_maker = pi_maker;
        Q_maker = q_maker;
        this.gamma = gamma;
    }
}

File: src/main/kotlin/lab/mars/rl/MDP.kt
package lab.mars.rl

/**
 * <p>
 * Created on 2017-08-31.
 * </p>
 *
 * @author wumo
 */
typealias StateSet = IndexedCollection<State?>
typealias StateValueFunction = IndexedCollection<Double>
typealias ActionValueFunction = IndexedCollection<Double>
typealias DeterminedPolicy = IndexedCollection<Action?>

class MDP(
        val states: StateSet,
        val gamma: Double,
        val v_maker: () -> StateValueFunction,
        val q_maker: () -> ActionValueFunction,
        val pi_maker: () -> DeterminedPolicy)

interface Indexable {
    val idx: IntArray
}

interface IndexedCollection<E> : Iterable<E> {
    operator fun get(vararg index: Int): E
    operator fun get(indexable: Indexable): E
    operator fun get(vararg indexable: Indexable): E
    operator fun set(vararg index: Int, s: E)
    operator fun set(indexable: Indexable, s: E)
    operator fun set(vararg indexable: Indexable, s: E)
}

class State(vararg index: Int) : Indexable {
    override val idx = index

    var actions: IndexedCollection<Action>? = null

    override fun toString() = idx.asList().toString()
}

class Action(vararg index: Int) : Indexable {
    override val idx = index

    var possibles: IndexedCollection<Possible>? = null

    fun sample(): Possible? = null

    override fun toString() = idx.asList().toString()
}

class Possible {
    var state: State? = null
    var reward: Double = 0.0
    var probability: Double = 0.0

    constructor()

    constructor(state: State, reward: Double, probability: Double) {
        this.state = state
        this.reward = reward
        this.probability = probability
    }
}
